### DISCUSSION TITLE
DataSource set as readOnly still allows mutable transactions

### DISCUSSION THREAD

**Discussion by Nahuel92** [micronaut-projects/micronaut-data#3104] [General]
Hi guys!

I ran into a very odd issue that I would like to share to see if it's my mistake, or if the framework is not intended for the way I'm using it.

I'm conditionally setting a `DataSource` as readOnly or not based on the value of a few properties via a `BeanCreatedEventListener<DatasourceConfiguration>` listener (thanks Graeme for this suggestion).

This works really well for single-threaded code, but when using multiple threads I noticed that this configuration is not honored, and write transactions are executed and the `readOnly` flag is silently ignored.

In any case, I created a PoC app that you can find below. This application sets the `readOnly` flag to true, to expose this behavior I described above:

https://github.com/Nahuel92/poc-projects/tree/master/micronaut/micronaut-data-multithreading-bug

I also did an ad hoc test with a `BeanCreatedEventListener<TransactionManagerConfiguration>` listener to set `enforceReadOnly` to true (in addition to setting the `readOnly` flag previously mentioned), but without any luck.

If more information is needed, please let me know.

I'll appreciate any clarification/help on this, thanks in advance!

**Comment by dstepanov** [comment]
It's probably database / driver related. Check if your data are saved.

**Reply by Nahuel92** [reply]
Hi dstepanov!

Yes, data is saved. I verified this with Postgres. I pushed a new test that passes when it should fail, and it verifies that the data is saved as if the DataSource wasn't set to readOnly mode.

**Reply by dstepanov** [reply]
It would be better to try some basic JDBC. without any Micronaut

**Reply by Nahuel92** [reply]
Thanks. I checked and the culprit is the Postgres driver. It has a not-so-clear behavior when `readOnly` is set.
Either `autoCommit` needs to be disabled, or `readOnlyMode` should be set to `always`:

https://github.com/pgjdbc/pgjdbc/blob/master/pgjdbc/src/main/java/org/postgresql/PGProperty.java#L529

Thanks a lot!
